security/rbac/engine/engine.go

```diff
@@ -0,0 +1,127 @@
+ * limitations under the License.
+ */
+
+package engine
+
+import (
+	"github.com/envoyproxy/go-control-plane/envoy/config/rbac/v2"
+	"google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+)
+
+type AuthorizationArgs struct {
+	md 			metadata.MD
+	peerInfo 	*peer.Peer
+}
+
+// Decision is the enum type that represents different authorization
+// decisions a CEL engine can return.
+type Decision int32
+
+const (
+	// DECISION_ALLOW indicates allowing the RPC to go through.
+	DECISION_ALLOW Decision = iota
+	// DECISION_DENY indicates denying the RPC from going through.
+	DECISION_DENY
+	// DECISION_UNKNOWN indicates that there is insufficient information to
+	// determine whether or not an RPC call is authorized.
+	DECISION_UNKNOWN
+)
+
+// Output of CEL engine.
+type AuthorizationDecision struct {
+	decision 				Decision
+	authorizationContext	string
+}
+
+// Action is the enum type that represents different actions that RBAC
+// policies can specify, in the case of a matching context.
+type Action int32
+
+const (
+	// ACTION_ALLOW indicates that if a policy is matched, authorization
+	// can be granted.
+	ACTION_ALLOW Action = iota
+	// ACTION_DENY indicates that if a policy is matched, authorization
+	// will be denied.
+	ACTION_DENY
+)
+
+// Returns whether or not a policy is matched.
+func matches(condition *expr.Expr, args AuthorizationArgs) bool {
+	// TODO
+	return false
+}
+
+// Struct for CEL engine.
+type CelEvaluationEngine struct {
+	action		Action
+	conditions	map[string]*expr.Expr
+}
+
+// Builds a CEL evaluation engine from Envoy RBAC.
+func RbacToCelEvaluationEngine(rbac envoy_config_rbac_v2.RBAC) CelEvaluationEngine {
+	var action Action
+	if rbac.Action == envoy_config_rbac_v2.RBAC_ALLOW {
+		action = ACTION_ALLOW
+	} else {
+		action = ACTION_DENY
+	}
+	
+	conditions := make(map[string]*expr.Expr)
+	for policyName, policy := range rbac.Policies {
+        conditions[policyName] = policy.Condition
+	}
+	
+	return CelEvaluationEngine{action, conditions}
+}
+
+// Builds a CEL evaluation engine from runtime policy template.
+// Currently do not have access to RTPolicyTemplate/can't import
+// func RTPolicyTemplateToCelEvaluationEngine(rt_policy RTPolicyTemplate) CelEvaluationEngine {
+// 	// TODO
+// 	return nil
+// }
+
+// The core function that evaluates whether an RPC is authorized.
+func (engine CelEvaluationEngine) evaluate(args AuthorizationArgs) AuthorizationDecision {
+	for policyName, condition := range engine.conditions {
+		if matches(condition, args) {
+			var decision Decision
+			if engine.action == ACTION_ALLOW {
+				decision = DECISION_ALLOW
+			} else {
+				decision = DECISION_DENY
+			}
+			return AuthorizationDecision{decision, "Policy matched: " + policyName}
+		}
+	}
+	// if no conditions matched
+	return AuthorizationDecision{DECISION_UNKNOWN, "No policies matched"}
+}
+
+func evaluate(args AuthorizationArgs) bool {
```

Comment:
Why do we need this function? In the main function, the user could directly New the RBAC engine, and then initiate the interceptor, and while inside the interceptor, they can use the RBAC engine instance to do the evaluation, right?

Reply:
Yes, but that would require us to let the user to write the interceptor based on where and how they make a new RBAC engine, which currently isn't the case. Do you want me to change it so there's a blank area in the interceptor functions where users need to fill out?

Reply:
The integration part of the interceptor is TBD. We MAY have a chain of RABC engines, and in the interceptor we will go through these engines one by one. This part is still not settled down yet, so even if you have an interceptor implementation ready, we are likely to change it later.
The reason why I asked you to take a look at the interceptor implementation is to give you a sense where the data is originated from, not to actually implement it. Having said that, in your main function, you can create a sample interceptor with only one RBAC engine inside, to illustrate how your code would work.

Reply:
Ok! I'll frame the entirety of interceptor code as part of the example then.